### PR TITLE
[BACKLOG-40519] Display the database connection in the UI based on th…

### PR DESCRIPTION
**Pull request by pdesai16**
…e level - Glabal, project and local. Changes include

 - Added SharedObjectIO that is responsible for reading/writing shared objects
 - Added the accessor method to get SharedObjectIO in Bowl. DefaultBowl provides the SharedOjectsIO for accessing global Shared objects.
 - Added DatabaseConnectionManager that uses the SharedObjectsIO to retrieve the shared objects
 - Updated DBConnectionFolderProvider to display the database connection based on the category - Project, Global and local.
 - Added a new method in  AbstractMeta to get the list of local datbase connection (that is defined n ktr/kjb)
 - Updated the unit tests

### CODE REVIEW

File: core/src/test/java/org/pentaho/di/shared/VfsSharedObjectsIOTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    // close the file
+    sharedFile.close();
+  }
+  private void validateNode (Node node ){
+    assertEquals("localhost", XMLHandler.getTagValue(node, "server"));
+    assertEquals("POSTGRESQL", XMLHandler.getTagValue(node, "type"));
+    assertEquals("Native", XMLHandler.getTagValue(node, "access"));
+    assertEquals("5435", XMLHandler.getTagValue(node, "port"));
+    assertEquals("postgres", XMLHandler.getTagValue(node, "username"));
```
Comment (abryant-hv) [inline core/src/test/java/org/pentaho/di/shared/VfsSharedObjectsIOTest.java:56]: formatting

File: engine/src/main/java/org/pentaho/di/base/AbstractMeta.java [hunk 2]
```diff
@@ -2175,4 +2174,7 @@ public void setMetaFileCache( IMetaFileCache metaFileCache ) {
     this.metaFileCache = metaFileCache;
   }
 
+  public List<DatabaseMeta> getLocalDbMetas() {
+    return new ArrayList<>( localDbMetas.values());
+  }
```
Comment (abryant-hv) [inline engine/src/main/java/org/pentaho/di/base/AbstractMeta.java:2179]: don't the callers modify this? (which is also not recommended)

File: ui/src/main/java/org/pentaho/di/ui/spoon/DatabasesCollector.java [hunk 3]
```diff
@@ -37,52 +38,74 @@
...
-    return names2metas.get( dbName );
+    return dbMetaMap.get( dbName );
   }
 
   private static IllegalStateException exception() {
     return new IllegalStateException( "Call collectDatabases() first" );
   }
+
+  private void addToMetaMap( List<DatabaseMeta> metaList) {
+    for (DatabaseMeta db : metaList) {
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/DatabasesCollector.java:107]: formatting.

File: ui/src/main/java/org/pentaho/di/ui/spoon/DatabasesCollector.java [hunk 4]
```diff
@@ -37,52 +38,74 @@
...
+
+    // SharedObjects
+    if ( dbManager != null ) {
+      dbMetaList = dbManager.getDatabases();
+      addToMetaMap( dbMetaList );
     }
 
+    // local databases
+    if ( meta != null ){
+      dbMetaList =meta.getLocalDbMetas();
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/DatabasesCollector.java:69]: formatting. 
Also weird to try to directly modify memory from another class.

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/provider/DBConnectionFolderProvider.java [hunk 5]
```diff
@@ -58,41 +64,59 @@ public DBConnectionFolderProvider() {
...
+        for ( String databaseName : collector.getDatabaseNames() ) {
+          if ( !filterMatch( databaseName, filter ) ) {
+            continue;
+          }
+          DatabaseMeta databaseMeta = collector.getMetaFor( databaseName );
+          projectDbNames.add( databaseMeta.getDisplayName() );
+          TreeNode childTreeNode = createTreeNode( treeNode, databaseMeta.getDisplayName(), guiResource
+            .getImageConnectionTree(), LeveledTreeNode.LEVEL.PROJECT, false );
+          if ( databaseMeta.isShared() ) {
+            childTreeNode.setFont( guiResource.getFontBold() );
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/tree/provider/DBConnectionFolderProvider.java:84]: isn't shared going away? why are we still bolding "shared" connections?

File: ui/src/test/java/org/pentaho/di/ui/spoon/SpoonRefreshDbConnectionsSubtreeTest.java [hunk 6]
```diff
@@ -22,21 +22,24 @@
 
 package org.pentaho.di.ui.spoon;
 
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.Test;
+import org.junit.*;
```
Comment (abryant-hv) [inline ui/src/test/java/org/pentaho/di/ui/spoon/SpoonRefreshDbConnectionsSubtreeTest.java:25]: don't import .*

**Review by abryant-hv — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/pentaho/di/core/bowl/DefaultBowl.java [hunk 7]
```diff
@@ -69,5 +74,12 @@ public void setMetastoreSupplier( Supplier<IMetaStore> metastoreSupplier ) {
     this.customSupplier = true;
   }
 
+  @Override
+  public SharedObjectsIO getSharedObjectsIO()  {
+    if ( sharedObjectsIO == null ) {
+      sharedObjectsIO = new VfsSharedObjectsIO("");
```
[the remark this answers is not in the document]
Comment (pdesai16) [reply core/src/main/java/org/pentaho/di/core/bowl/DefaultBowl.java:80]: The VfsSharedObjectIO applies the rules like reading the kettle environment variabe and f not read the  user's home directory

File: core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+   * Loads the shared objects in the map. The map will be of the form <String, Node>
+   * where key can be {"connection", "slaveserver", "partitionschema" or  clusterschema"} and
+   * value will be xml Node.
+   *
+   * @param rootFolder
+   * @throws KettleXMLException
+   */
+  public void loadSharedObjectNodeMap( String rootFolder ) throws KettleXMLException {
+    // Get the complete path to shared.xml
+    this.sharedObjectFile = getSharedObjectFilePath( rootFolder );
```
[the remark this answers is not in the document]
Comment (pdesai16) [reply core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java:101]: If there are multiple callers of this code, they all will have to define the same rules on their side. It is easy to centralize it so caller does not care

File: core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+  public void saveSharedObject( String type, String name, Node node ) throws IOException, KettleException {
+    // Find the map for the type
+    Map<String, Node> nodeMap = getNodesMapForType( type );
+    if ( nodeMap.containsKey( name ) ) {
+      // Update the map entry for this name
+      nodeMap.put( name, node );
+
+      FileObject fileObject = KettleVFS.getFileObject( sharedObjectFile );
+      String backupFileName = createOrGetFileBackup( fileObject );
+      writeToFile( fileObject, backupFileName );
```
[the remark this answers is not in the document]
Comment (pdesai16) [reply core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java:179]: We should wait till we work on the right click menu items. I haven't tested this part at all

File: core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+   * @param type The SharedObject type
+   * @param name The name of the SharedObject for a given type
+   * @param node The xml node that contains the details of the SharedObject
+   * @throws IOException
+   * @throws KettleException
+   */
+  public void saveSharedObject( String type, String name, Node node ) throws IOException, KettleException {
+    // Find the map for the type
+    Map<String, Node> nodeMap = getNodesMapForType( type );
+    if ( nodeMap.containsKey( name ) ) {
```
[the remark this answers is not in the document]
Comment (abryant-hv) [reply core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java:173]: Maybe add a TODO?

File: core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+    Node removedNode = nodeTypeMap.remove( name );
+  }
+
+  private Map<String, Node> getNodesMapForType( String type ) throws KettleXMLException {
+    // change the type to uppercase so we can convert it to SharedObjectType
+    return getNodesMapForType( SharedObjectType.valueOf( type.toUpperCase() ) );
+  }
+
+  private Map<String, Node> getNodesMapForType( SharedObjectType type ) throws KettleXMLException {
+    Map<String, Node> nodeMap = new HashMap();
```
[the remark this answers is not in the document]
Comment (pdesai16) [reply core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java:286]: The map is assigned based on the type in the switch

[the remark this answers is not in the document]
Comment (abryant-hv) [reply core/src/main/java/org/pentaho/di/shared/VfsSharedObjectsIO.java:286]: oh, it gets returned. I think I missed the return before? or something.

File: engine/src/main/java/org/pentaho/di/base/AbstractMeta.java (hunk 2 above)
Comment (pdesai16) [reply engine/src/main/java/org/pentaho/di/base/AbstractMeta.java:2174]: localDbMetas is something I added. There is currently no setters for it. It gets loaded during loadXml(). We will need to handle the case when someone is updating the ktr/kjb or creating a new ktr/kjb.

**Review by abryant-hv — APPROVED** [review]

File: engine/src/main/java/org/pentaho/di/core/KettleEnvironment.java [hunk 12]
```diff
@@ -256,4 +261,12 @@ public static void reset() {
     LoggingRegistry.getInstance().reset();
     initialized.set( null );
   }
+
+  /**
+   * Registers the SharedObject type specific factory with the Bowl factory registry.
+   */
+  private static void registerSharedObjectManagersWithBowl(){
```
Comment (tgf) [inline engine/src/main/java/org/pentaho/di/core/KettleEnvironment.java:268]: formatting

File: engine/src/main/java/org/pentaho/di/job/JobMeta.java [hunk 13]
```diff
@@ -1013,6 +1013,7 @@ public void loadXML( Node jobnode, String fname, Repository rep, IMetaStore meta
         dbcon.shareVariablesWith( this );
         if ( !dbcon.isShared() ) {
           privateDatabases.add( dbcon.getName() );
+          localDbMetas.put(dbcon.getName(), dbcon );
```
Comment (tgf) [inline engine/src/main/java/org/pentaho/di/job/JobMeta.java:1016]: formatting

File: engine/src/main/java/org/pentaho/di/shared/DatabaseConnectionManager.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+  }
+  private DatabaseConnectionManager( Bowl bowl ) throws KettleXMLException {
+    this.bowl = bowl;
+    this.sharedObjectIO = bowl.getSharedObjectsIO();
+  }
+
+
+  public List<DatabaseMeta> getDatabases() throws KettleXMLException {
+    Map<String, Node> nodeMap = sharedObjectIO.getSharedObjects( String.valueOf( SharedObjectsIO.SharedObjectType.CONNECTION ) );
+    populateDbMetaMap ( nodeMap );
```
Comment (tgf) [inline engine/src/main/java/org/pentaho/di/shared/DatabaseConnectionManager.java:53]: formatting

File: engine/src/main/java/org/pentaho/di/shared/DatabaseConnectionManager.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+  }
+
+  /**
+   * Factory for the DatabaseConnectionManager. This factory class is registered with BowlFactory registry
+   * during the initialization in KettleEnvironment
+   */
+  public static class DbConnectionManagerFactory implements ManagerFactory<DatabaseConnectionManager> {
+    public DatabaseConnectionManager apply( Bowl bowl ) throws KettleException {
+      DatabaseConnectionManager dbConnectionManager = DatabaseConnectionManager.getInstance( bowl );
```
Comment (tgf) [inline engine/src/main/java/org/pentaho/di/shared/DatabaseConnectionManager.java:77]: Could just return here. Think sonar will complain about it

File: engine/src/main/java/org/pentaho/di/shared/DelegatingShareObjectsIO.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+public class DelegatingShareObjectsIO implements SharedObjectsIO {
+
+  List<SharedObjectsIO> sharedObjectsIOList;
+  public DelegatingShareObjectsIO( SharedObjectsIO... sharedObjectsIOs ) {
+    this.sharedObjectsIOList = new ArrayList<>(Arrays.asList( sharedObjectsIOs ) );
+  }
+
+  @Override
+  public Map<String, Node> getSharedObjects(String type) throws KettleXMLException {
```
Comment (tgf) [inline engine/src/main/java/org/pentaho/di/shared/DelegatingShareObjectsIO.java:39]: formatting (also last 2 methods)

**Review by tgf — APPROVED** [review]

**Pull request merged**